### PR TITLE
Sort the address book by name and mail address

### LINKED ISSUE #402: Sorting the address book
As a user I want to sort my address book entries to have a better overview over my contacts​

**acceptance criteria:​**

- The address book dialog window has a Sort Panel ​
- Sort Panel consists of sort-by-name, sort-by-mail button​
- Sort-by-name button action: sorts the address book entries by name​
- sort-by-mail button action: sorts the address bool entries by mail​

time needed: M

### PR DESCRIPTION
**Pull request by laugengebaeck**
This closes #402.

UI of the ICAdressBookDialog now looks like this:

addressbook_sort

### CODE REVIEW

File: packages/IMAPClient-Core.package/ICEmail.class/instance/senderName.st [hunk 1]
```diff
@@ -3,4 +3,4 @@ senderName
 
 	| fromCollection |
 	fromCollection := self from findBetweenSubStrs: self class senderAddressDelimeters.
-	^ (fromCollection at: 1 ifAbsent: [ '' ])
\ No newline at end of file
+	^ ((fromCollection at: 1 ifAbsent: [ '' ]) subStrings: '"') at: 1
```
Comment (elena-sys) [inline packages/IMAPClient-Core.package/ICEmail.class/instance/senderName.st:6]: Changing the default value to an empty string is a different user story. Please remove this change

File: packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/sortMethod.st [hunk 2]
```diff
@@ -0,0 +1,3 @@
+accessing
+sortMethod
```
Comment (elenagensch) [inline packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/sortMethod.st:2]: newline between name and body

File: packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/sortMethod..st [hunk 3]
```diff
@@ -0,0 +1,3 @@
+accessing
+sortMethod: anObject
```
Comment (elenagensch) [inline packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/sortMethod..st:2]: newline between name and body

File: packages/IMAPClient-Core.package/ICEmail.class/instance/senderName.st (hunk 1 above)
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICEmail.class/instance/senderName.st:6]: You maybe misunderstood what this does. This removes the quotation marks (which are sometimes included in the sender name) to make a consistent sorting order possible.

File: packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/sortContactsBy..st [hunk 4]
```diff
@@ -0,0 +1,8 @@
+action
+sortContactsBy: aMethod
+	
+	(self sortMethod = aMethod) ifTrue: [self reverseContacts] ifFalse: [
```
Comment (sm1lla) [inline packages/IMAPClient-UI.package/ICAddressBookDialog.class/instance/sortContactsBy..st]: No brackets around `self sortedMethod = aMethod` needed. Also, maybe put the `ifFalse` statement in a new line

**Review by sm1lla — CHANGES_REQUESTED** [review]
Looks good, but please add tests.

**Comment by laugengebaeck** [comment]
We should maybe also invalidate the sorting when a new contact is added by the user, because otherwise this new contact might not be sorted into the right order.

Edit: This is done now.

**Review by blackjack2693 — APPROVED** [review]

**Comment by laugengebaeck** [comment]
Another thing we should think about before merging this pull request (we talked about that last week): Should we change the address book to use ICContacts internally? The current sorting method is quite ugly (and hard to understand) because it works by splitting strings.

**Comment by elenagensch** [comment]
Yeah, this is indeed ugly. We decided to pass Contacts as Strings in order to use `addAllFirstUnlessAlreadyPresent:` (https://github.com/hpi-swa-teaching/IMAPClient/blob/develop/packages/IMAPClient-Core.package/ICAddressBook.class/instance/addEmail..st#L4). When they are passed as ICContacts every time you open the address book you attach all contacts to the existing address book (because their id is unique). But that decision was made before sorting, so dissolving that makes sense.

**Comment by sm1lla** [comment]
If that makes it simpler we should definitly change it. But then this should be done this week, as we are planning to merge on Monday at the lastest.

**Comment by laugengebaeck** [comment]
The tests requested by @sm1lla should be added now. 

I decided not to change the `ICAddressBook` to use `ICContact`s since the `PluggableTreeSpec` we use for displaying can only display the collection it is watching directly (without `asString`). We could use a `PluggableMultiColumnListSpec` or something like that (which would resolve that issue), but this seems to require a lot of complicated changes.

**Review by sm1lla — APPROVED** [review]

**Comment by laugengebaeck** [comment]
Can I merge this? (I would list Christian as the first author)

**Pull request merged**